src/tdycharacteristiccurves.c

```diff
@@ -231,6 +274,149 @@ void RelativePermeability_Mualem(PetscReal m,PetscReal Se,PetscReal *Kr,
   }
 }
 
+/* -------------------------------------------------------------------------- */
+///
+/// Sets up a cubic polynomial interpolation for relative permeability following
+/// PFLOTRAN's approach of smoothing relative permeability functions
+///
+/// @param [in] x1           low value of x
+/// @param [in] x2           high value of x
+/// @param [inout] *rhs      rhs vector (input) and coefficients (output)
+///
+///  f(x) = a0 + a1 * x + a2 * x^2 + a2 * x^3
+///  df_dx = a1 + 2 * a2 * x + 3 * a3 * x^2
+///
+/// Constraints:
+/// f(low)      = rel_perm_fn(x)      = f1
+/// f(high)     = 1.0                 = f2
+/// df_dx(low)  = drv_rel_perm_fn(x)  = df1_dx
+/// df_dy(high) = 0.0                 = df2_dx
+///
+/// Linear system:
+/// a0 + a1 * x1 + a2 * x1^2   + a2 * x1^3     = f1
+/// a0 + a1 * x2 + a2 * x1^2   + a2 * x2^3     = f2
+///      a1      + 2 * a2 * x1 + 3 * a3 * x1^2 = df1_dx
+///      a1      + 2 * a2 * x2 + 3 * a3 * x2^2 = df2_dx
+///
+PetscErrorCode CubicPolynomialSetup(PetscReal x1, PetscReal x2, PetscReal *rhs) {
+
+  PetscInt n = 4, nrhs = 1;
+
+  PetscReal A[16] = {
+     1.0,                  1.0,                  0.0,                      0.0,
+     x2,                   x1,                   1.0,                      1.0,
+     PetscPowReal(x2,2.0), PetscPowReal(x1,2.0), 2.0*x2,                   2.0*x1,
+     PetscPowReal(x2,3.0), PetscPowReal(x1,3.0), 3.0*PetscPowReal(x2,2.0), 3.0*PetscPowReal(x1,2.0)
+  };
+
+   PetscInt lda=4, ldb=4;
+   PetscInt info; // success/failure from LAPACK
+   PetscInt ipiv[n]; // pivot indices
+
+   dgesv_( &n, &nrhs, A, &lda, ipiv, rhs, &ldb, &info );
+
+  PetscFunctionReturn(0);
+}
+
+/* -------------------------------------------------------------------------- */
+///
+/// Computes the value and derivative of the cubic polynomial
+///
+/// @param [in] *coeffs  coefficients of a cubic polynomals
```

Comment:
We don't normally include the asterisk in documented parameter names. C pointer declarations are weird and confusing--the variable name is `x`, and its type is `PetscReal *`, but folks customarily stick the asterisk next to the variable name to allow the declaration of multiple pointer variables.

For example, suppose you wanted pointers to `PetscReal` values named `x`, `y`, and `z`. One might think that since the type you want is `PetscReal *`, you would say something like

```
PetscReal* x, y, z;
```

This is wrong--this gives you a pointer `x` and two `PetscReal`s `y` and `z`. The right way to declare the three pointers is

```
PetscReal *x, *y, *z;
```

It's one of the most confusing language features.

Reply:
Even more unfortunate is that one "dereferences" a pointer by sticking an asterisk in front of it:

```
*x = 1; // set x to 1
```

So now it's possible for someone to think that `*x` is how you refer to the pointer. The pointer is `x`, and `*x` is the value that it points to. Between that and the type declaration, there's a whole lot of confusion out there.